### ISSUE TITLE
Proposal: Support Decentralized Identifiers (DIDs) in addition to Web IDs

### ISSUE THREAD

**Issue by dmitrizagidulin** [solid/identity-panel#1]
This is a placeholder proposal for the Solid project to support Decentralized Identifiers (DIDs), in addition to Web IDs.

![DIDs vs WebIDs](https://user-images.githubusercontent.com/947005/65087622-6ed6eb80-d984-11e9-84fd-636d02b8ed36.png)

Similarities:

* Both are **URIs** (follow the URI RFC), both resolve to a Linked Data document
* Both are used as **identifiers** for individual people, organizations, services, and other entities
* Both are intended to **store Public Key material**, for uses such as authentication
* Both **store links** to documents or API endpoints (such as preferences, OIDC Issuers, etc)
* Essentially interoperable

Differences:

* **PII**: In addition to key material and links, Web ID Profiles also contain Personally Identifiable Information (PII) such as User Profile Info (name, email, user picture), or the user's Social Graph (using things like `foaf:knows`). DID Documents, on the other hand, are required (due to GDPR pressures and similar) to _not_ store any sort of PII, as much as is possible.
* **Vocab**: Web ID Profiles use a combination of `foaf`, `solid` and `cert` vocabularies. DID Documents use the `did` vocab.
* **Key Type Support**: DIDs use the Linked Data Proof and Linked Data Signatures specs, and support a wider array of key types (RSA, Ed25519, Curve25519, Secp256k1, etc) and purposes (authentication, digital signatures, encryption, delegation, capability invocation)
* **Storage and Resolution mechanisms**: There are many different DID methods; some of them store their DID Documents on distributed ledgers, some of them on the Web (see the [`did:web` method](https://github.com/w3c-ccg/did-method-web)), some on IPFS, some on IoT devices, etc.
* **W3C Level of Standardization**: The [DID spec](https://github.com/w3c-ccg/did-spec) has been incubated for 3 years in the [Credentials Community Group](https://www.w3.org/community/credentials/). Recently (as of this TPAC happening this week), an official W3C DID Working Group has been chartered to continue standardization.

(Inrupt, as well as the Solid community, should join the DID WG!)

**Comment by justinwb** [comment]
Fully support incorporating DID into the Solid ecosystem!

**Comment by creatinglake** [comment]
I fully support this effort as well! Solid and DIDs are synergistic.

**Comment by RubenVerborgh** [comment]
Yes, but which resolvers, and how will this work in the browser?

I’m all for DID, but what is the subset implementers MUST support? That is _the_ issue.

**Comment by csarven** [comment]
What's the essential difference between `did:web:` and `http:` ? What's possible with `did:web:` that's not possible with `http:`?

[IIRC, in order to join the WG, one would need to be an IE or participate on behalf of a Member org.]

**Comment by csarven** [comment]
`did:web:` isn't yet a registered method but I'm sure that can change easily.

A DID document is represented in JSON-LD and its context uses an HTTP URI ( eg https://w3id.org/did/v1 ) ... So, HTTP all the way down?

**Comment by dmitrizagidulin** [comment]
@RubenVerborgh 
> Yes, but which resolvers, and how will this work in the browser?
> I’m all for DID, but what is the subset implementers MUST support? That is the issue.

Excellent questions. So, the specs do not, at the moment, specify any required subset.
You're right, that'll be one of our main decisions - which DID methods to support?

I'd advocate for starting with the `did:web` method (since it aligns the closest to the Solid ecosystem), and then add support for others as time/resources/usecase permits. (And, all the required libraries do work in the browser.)

**Comment by dmitrizagidulin** [comment]
@csarven 
> What's possible with `did:web:` that's not possible with `http:`?

The way I figure it, the important part is the data model / shape of the DID Document. It could either be linked to via an `https` or a `did:web` link.

**Comment by RubenVerborgh** [comment]
> then add support for others as time/resources/usecase permits

But it’s not about our resources though. It’s about what we put as a MUST in the spec.

Only those DID resolvers with a MUST can ever be chosen safely for a WebID. All the others risk being unsupported by any given app.

**Comment by bblfish** [comment]
- Would the [User controlled Authorization App and App launcher proposal](https://github.com/solid/authorization-and-access-control-panel/issues/45) help for DIDs?
-  Can DIDs be used with HTTP Signature? see [issue regarding get HTTP-Sig to work with Solid using https keyIds](https://github.com/solid/authentication-panel/issues/18) or the [HTTP Sig proposal for Solid PR](https://github.com/w3c-dvcg/http-signatures/issues/92)

**Comment by dmitrizagidulin** [comment]
(Reference / further info) There was an excellent introductory presentation on DIDs on this week's Credentials Community Group call that might also be useful to this community: 

[Decentralized Identifiers (DIDs) & Self-Sovereign Identity (SSI): A Non Technical Discussion](https://docs.google.com/presentation/d/1O_jjvRtTsmFSucuw3LdtaVewCRMiX3DmmajT_h20a4w)

**Comment by csarven** [comment]
`did:web` may be in conflict with the following criteria from the DID spec ( https://www.w3.org/TR/2019/WD-did-core-20191209/#did-method-schemes ):

>The `method-specific-id` value MUST be able to be generated without the use of a centralized registry service.

in that domain names and IP address assignments are handled by ICANN and IANA. Putting the challenge around resolving aside, DID's notion of "centralized" and the requirements thereof is particularly important if http-based WebIDs are intended to (or can ever) work under did:web.

**Comment by RubenVerborgh** [comment]
> The `method-specific-id` value MUST be able to be generated without the use of a centralized registry service.

I don't see an objection in the way that it is currently phrased, in the sense that DNS is _able_ to be realized in a non-centralized way. It just so happens that most implementations today are rather centralized.
